xrpl/transaction/main.py

```diff
@@ -79,7 +105,83 @@ def submit_transaction(
+    transaction: Transaction,
+    wallet: Wallet,
+) -> Dict[str, Any]:
+    """
+    Prepares a Transaction by converting it to a JSON-like dictionary, converting the
+    field names to CamelCase. If a Client is provided, then it also autofills any
+    relevant fields.
+
+    Args:
+        transaction: the Transaction to be prepared.
+        wallet: the wallet that will be used for signing.
+
+    Returns:
+        A JSON-like dictionary that is ready to be signed.
+
+    Raises:
+        XRPLException: if both LastLedgerSequence and `ledger_offset` are provided, or
+            if an address tag is provided that does not match the X-Address tag.
+    """
+    transaction_json = transaction_json_to_binary_codec_form(transaction.to_dict())
+    transaction_json["SigningPubKey"] = wallet.pub_key
+
+    _validate_account_xaddress(transaction_json, "Account", "SourceTag")
+    if "Destination" in transaction_json:
+        _validate_account_xaddress(transaction_json, "Destination", "DestinationTag")
+
+    # DepositPreauth
+    _convert_to_classic_address(transaction_json, "Authorize")
+    _convert_to_classic_address(transaction_json, "Unauthorize")
+    # EscrowCancel, EscrowFinish
+    _convert_to_classic_address(transaction_json, "Owner")
+    # SetRegularKey
+    _convert_to_classic_address(transaction_json, "RegularKey")
+
+    return transaction_json
+
+
+def _autofill_transaction(transaction: Transaction, client: Client) -> Transaction:
+    transaction_json = transaction.to_dict()
+    if "sequence" not in transaction_json:
+        sequence = get_next_valid_seq_number(transaction_json["account"], client)
+        transaction_json["sequence"] = sequence
+    if "fee" not in transaction_json:
+        transaction_json["fee"] = get_fee(client)
+
+    if "last_ledger_sequence" not in transaction_json:
+        ledger_sequence = get_latest_validated_ledger_sequence(client)
+        transaction_json["last_ledger_sequence"] = ledger_sequence + _LEDGER_OFFSET
+    return Transaction.from_dict(transaction_json)
+
+
+def _validate_account_xaddress(
+    json: Dict[str, Any], account_field: str, tag_field: str
+) -> None:
+    """
+    Mutates JSON-like dictionary so the X-Address in the account field is the classic
+    address, and the tag is in the tag field.
+    """
+    if is_valid_xaddress(json[account_field]):
+        account, tag, _ = xaddress_to_classic_address(json[account_field])
+        json[account_field] = account
+        if json[tag_field] and json[tag_field] != tag:
+            raise XRPLException(f"{tag_field} value does not match X-Address tag")
+        json[tag_field] = tag
+
+
+def _convert_to_classic_address(json: Dict[str, Any], field: str) -> None:
+    """
+    Mutates JSON-like dictionary to convert the given field from an X-address (if
+    applicable) to a classic address.
+    """
+    if field in json and is_valid_xaddress(json[field]):
+        json[field] = xaddress_to_classic_address(json[field])
```

Comment:
Should we raise an exception if the user provides an X-address with a tag to a field that doesn't relate to tags? I'm honestly not sure, but it feels weird to use an X-address with a tag for `Authorize` for example when it would really authorize anything that shares the same underlying classic address.

Reply:
I agree it's weird, but I could see the case where you use an X-Address for everything so you just stick it in here too.

Reply:
Using an X-address with a tag and without a tag has different implications. (This gave me an interesting idea for how Deposit Preauthorization could be limited to a specific source tag, but it can't work that way currently.) So it might be surprising for users who think they've authorized only "Alice at Bitstamp" when really they've authorized "Bitstamp" so anyone could send to them from there.

Doesn't need to change now, but worth keeping an eye on.